### ISSUE TITLE
BSOD on IPFS (golang)

### ISSUE THREAD

**Issue by ProIcons** [Microsoft/WSL#2669]
So lately i'm experimenting with IPFS which it's first implementation is written in go. So during the past 5 days i encountered many crahses only on IPFS.

Using Ubuntu distro, both 14 and 16 versions, you may encounter bsod on ipfs daemon
and you will encounter certainly bsod on ipfs-update.

To reproduce
you can download the binary from their site https://dist.ipfs.io/#ipfs-update

```
wget https://dist.ipfs.io/ipfs-update/v1.5.2/ipfs-update_v1.5.2_linux-amd64.tar.gz
tar -xzf ipfs-update_v1.5.2_linux-amd64.tar.gz
cd ipfs-update
./install.sh
ipfs-update install latest
```

while it will say "verifying..." which points to https://github.com/ipfs/ipfs-update/blob/master/lib/install.go#L85
which i guess points to that function:
https://github.com/ipfs/ipfs-update/blob/master/test-dist/testnew.go#L177
 , you will get a blue screen with bug check string "SYSTEM_SERVICE_EXCEPTION"

Microsoft Windows [Version 10.0.16299.64]

Minidump sent to secure@microsoft.com

**Comment by ProIcons** [comment]
```
3: kd> !analyze -v
*******************************************************************************
*                                                                             *
*                        Bugcheck Analysis                                    *
*                                                                             *
*******************************************************************************

SYSTEM_SERVICE_EXCEPTION (3b)
An exception happened while executing a system service routine.
Arguments:
Arg1: 00000000c0000005, Exception code that caused the bugcheck
Arg2: fffff8037db56a4e, Address of the instruction which caused the bugcheck
Arg3: ffffbb886f883870, Address of the context record for the exception that caused the bugcheck
Arg4: 0000000000000000, zero.

Debugging Details:
------------------


DUMP_CLASS: 1

DUMP_QUALIFIER: 400

BUILD_VERSION_STRING:  10.0.16299.64 (WinBuild.160101.0800)

DUMP_TYPE:  2

DUMP_FILE_ATTRIBUTES: 0xc
  Insufficient Dumpfile Size
  Kernel Generated Triage Dump

BUGCHECK_P1: c0000005

BUGCHECK_P2: fffff8037db56a4e

BUGCHECK_P3: ffffbb886f883870

BUGCHECK_P4: 0

EXCEPTION_CODE: (NTSTATUS) 0xc0000005 - The instruction at 0x%p referenced memory at 0x%p. The memory could not be %s.

FAULTING_IP: 
nt!ExpCheckForFreedEnhancedTimer+a
fffff803`7db56a4e 388198000000    cmp     byte ptr [rcx+98h],al

CONTEXT:  ffffbb886f883870 -- (.cxr 0xffffbb886f883870)
rax=000000018e9a93e2 rbx=0000000000000000 rcx=0000000000000000
rdx=0000000000000000 rsi=0000000000000000 rdi=0000000000000000
rip=fffff8037db56a4e rsp=ffffbb886f884260 rbp=fffffffffd050f80
 r8=0000000000000004  r9=0000000000000000 r10=7ffffffffffffffc
r11=ffffbb886f884388 r12=ffffbb886f884800 r13=0000000000000000
r14=ffff920900aa1000 r15=0000000000000000
iopl=0         nv up ei ng nz na po nc
cs=0010  ss=0018  ds=002b  es=002b  fs=0053  gs=002b             efl=00010286
nt!ExpCheckForFreedEnhancedTimer+0xa:
fffff803`7db56a4e 388198000000    cmp     byte ptr [rcx+98h],al ds:002b:00000000`00000098=??
Resetting default scope

CPU_COUNT: 4

CPU_MHZ: cef

CPU_VENDOR:  GenuineIntel

CPU_FAMILY: 6

CPU_MODEL: 2a

CPU_STEPPING: 7

CUSTOMER_CRASH_COUNT:  1

DEFAULT_BUCKET_ID:  WIN8_DRIVER_FAULT

BUGCHECK_STR:  0x3B

CURRENT_IRQL:  0

ANALYSIS_SESSION_HOST:  ATLANTIS

ANALYSIS_SESSION_TIME:  11-15-2017 02:48:56.0165

ANALYSIS_VERSION: 10.0.17016.1000 amd64fre

LAST_CONTROL_TRANSFER:  from fffff8037db56a15 to fffff8037db56a4e

STACK_TEXT:  
ffffbb88`6f884260 fffff803`7db56a15 : 00000000`00000000 00000000`00000002 ffffbb88`6f884448 fffff802`51837c7f : nt!ExpCheckForFreedEnhancedTimer+0xa
ffffbb88`6f8842a0 fffff802`5182e4db : 00000001`8e9a939c ffffbb88`6f884438 ffffbb88`6f8843c0 ffffbb88`6f884448 : nt!ExSetTimer+0x55
ffffbb88`6f8842f0 fffff802`51875f4e : 00000000`00000005 ffff9209`00000000 ffffbb88`6f8843f0 ffff9209`00ba6ea0 : LXCORE!LxpThreadWait+0xcb
ffffbb88`6f884390 fffff802`51876352 : ffff9209`00ba6e03 ffffa58b`5115cad0 00000005`00000001 ffffffff`fd050f80 : LXCORE!LxpSocketInetStreamApplyLinger+0x102
ffffbb88`6f884430 fffff802`51882f09 : ffff9209`00ba6ea0 ffff9208`e1f08570 ffffbb88`6f8848b0 ffff9208`e1f08570 : LXCORE!LxpSocketInetStreamFileReleaseInternal+0x8a
ffffbb88`6f884460 fffff802`51875513 : ffff9208`e25bf008 ffff9208`e1f08598 ffffa58b`00000feb fffff803`2020784c : LXCORE!LxpSocketInetStreamFileRelease+0x9
ffffbb88`6f884490 fffff802`51873126 : ffff9209`00ba6ea0 ffffbb88`6f8848b0 ffff9209`00bf00e0 ffff9208`e1f08570 : LXCORE!LxpSocketInetFileRelease+0x33
ffffbb88`6f8844c0 fffff802`51841c39 : 00000000`00000000 ffff9209`00bf00e0 ffffbb88`6f8848b0 ffffbb88`6f8848b0 : LXCORE!LxpSocketFileRelease+0x36
ffffbb88`6f8844f0 fffff802`5180bb17 : 00000000`00000000 00000000`00000000 ffff9209`00bf00f0 ffff9209`00bf00c0 : LXCORE!VfsFileNotifyFileDescriptorClose+0x51
ffffbb88`6f884520 fffff802`5180c0a4 : ffff9208`e1f08570 ffffbb88`6f884670 ffffbb88`6f8848b0 00000000`0000000f : LXCORE!LxpFileDescriptorCloseInternal+0x87
ffffbb88`6f884550 fffff802`5185b38d : ffff9208`ff8898d0 ffffbb88`6f8848b0 ffffa58b`4a0e7400 fffff803`7db2673c : LXCORE!LxpFileDescriptorTableDereference+0x54
ffffbb88`6f884580 fffff802`5185994f : 00000000`00000000 ffff9208`f34e4750 ffffa58b`4a0e7720 fffff803`7db268e9 : LXCORE!LxpProcFsFdDereferenceList+0xd
ffffbb88`6f8845b0 fffff802`51858e21 : ffff9208`ff8898d0 ffffbb88`6f8848b0 ffffbb88`6f8848b0 ffffbb88`6f884670 : LXCORE!LxpPseudoFsDynamicEntryDirectoryDelete+0x1f
ffffbb88`6f8845e0 fffff802`51843220 : ffff9208`f34e4700 ffffbb88`00000000 ffffbb88`00000000 ffffbb88`6f8848b0 : LXCORE!LxpPseudoFsDirectoryInodeDelete+0x41
ffffbb88`6f884610 fffff802`51858b4e : ffff9208`ff8898d0 ffffbb88`6f8848b0 ffff9208`fa358be0 ffffbb88`6f8848b0 : LXCORE!VfsInodeDereference+0x80
ffffbb88`6f884640 fffff802`5185a102 : ffff9209`009d0080 ffff9208`e1e58e70 ffffbb88`6f8848b0 ffffbb88`6f8848b0 : LXCORE!LxpPseudoFsDeleteDirectoryChildren+0x1be
ffffbb88`6f884680 fffff802`51859f9a : ffff9208`e1e58e58 ffff9208`f9d4a5a0 ffffbb88`6f8848b0 00000000`00000000 : LXCORE!LxpPseudoFsDynamicEntryRemoveEnd+0xd2
ffffbb88`6f8846b0 fffff802`517c2e2d : ffff9208`e1e58e98 00000000`00000000 ffff9208`f9d4a5a0 ffffbb88`6f8848b0 : LXCORE!LxpPseudoFsDynamicEntryRemove+0x4a
ffffbb88`6f8846e0 fffff802`5183027d : ffff9209`007bb6f0 ffffbb88`00000000 ffff9208`00000000 ffff9208`f9d4a5a0 : LXCORE!LxpProcFsThreadGroupCleanup+0x25
ffffbb88`6f884710 fffff802`5182c3cb : ffff9208`e25bf000 00007f49`00000000 00000000`00000000 ffffa58b`4a0e7400 : LXCORE!LxpThreadGroupEnd+0x541
ffffbb88`6f884810 fffff802`5182ddb4 : 00000000`00000011 ffffbb88`6f8848b0 ffff9209`00aa1000 ffffbb88`6f8849d8 : LXCORE!LxpThreadEnd+0x207
ffffbb88`6f884860 fffff802`5189bece : ffffa58b`509c1e10 ffffbb88`6f884988 ffffbb88`6f884b00 ffffbb88`6f8849d8 : LXCORE!LxpThreadSignalApc+0xb8
ffffbb88`6f884890 fffff803`7dae91c7 : ffffa58b`4a0e7400 000000c4`2005c810 ffff9209`00aa1000 00000000`000000ca : LXCORE!LxpNtStateThreadApcUser+0x3e
ffffbb88`6f884930 fffff803`7dbeb830 : 000000c4`2005c810 ffffbb88`6f8849c0 ffffffff`00000000 00000000`00000000 : nt!KiDeliverApc+0x337
ffffbb88`6f8849c0 fffff803`7dbf379e : ffffbb88`6f884b00 00000000`00000000 00000000`00000000 00000000`00000000 : nt!KiInitiateUserApc+0x70
ffffbb88`6f884b00 00000000`009e2bd3 : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`00000000 : nt!KiSystemServiceExitPico+0x47
00007f49`48b2fe10 00000000`00000000 : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`00000000 : 0x9e2bd3


THREAD_SHA1_HASH_MOD_FUNC:  b744a4779eb085ac2704d018f1d2170cdd44e8a0

THREAD_SHA1_HASH_MOD_FUNC_OFFSET:  8f06f3cd34ee50ad024756b03249fe212684bebd

THREAD_SHA1_HASH_MOD:  fb55a0bc34d9a6cf129a9e051ec1616771d1a5f4

FOLLOWUP_IP: 
LXCORE!LxpThreadWait+cb
fffff802`5182e4db 488b4c2448      mov     rcx,qword ptr [rsp+48h]

FAULT_INSTR_CODE:  244c8b48

SYMBOL_STACK_INDEX:  2

SYMBOL_NAME:  LXCORE!LxpThreadWait+cb

FOLLOWUP_NAME:  MachineOwner

MODULE_NAME: LXCORE

IMAGE_NAME:  LXCORE.SYS

DEBUG_FLR_IMAGE_TIMESTAMP:  0

IMAGE_VERSION:  10.0.16299.15

STACK_COMMAND:  .cxr 0xffffbb886f883870 ; kb

BUCKET_ID_FUNC_OFFSET:  cb

FAILURE_BUCKET_ID:  0x3B_LXCORE!LxpThreadWait

BUCKET_ID:  0x3B_LXCORE!LxpThreadWait

PRIMARY_PROBLEM_CLASS:  0x3B_LXCORE!LxpThreadWait

TARGET_TIME:  2017-11-15T00:34:57.000Z

OSBUILD:  16299

OSSERVICEPACK:  64

SERVICEPACK_NUMBER: 0

OS_REVISION: 0

SUITE_MASK:  272

PRODUCT_TYPE:  1

OSPLATFORM_TYPE:  x64

OSNAME:  Windows 10

OSEDITION:  Windows 10 WinNt TerminalServer SingleUserTS

OS_LOCALE:  

USER_LCID:  0

OSBUILD_TIMESTAMP:  2017-10-25 06:06:03

BUILDDATESTAMP_STR:  160101.0800

BUILDLAB_STR:  WinBuild

BUILDOSVER_STR:  10.0.16299.64

ANALYSIS_SESSION_ELAPSED_TIME:  5d7

ANALYSIS_SOURCE:  KM

FAILURE_ID_HASH_STRING:  km:0x3b_lxcore!lxpthreadwait

FAILURE_ID_HASH:  {6f7b88fb-a543-8bec-0913-72675f732cdc}

Followup:     MachineOwner
---------
```

**Comment by jstarks** [comment]
Thanks for submitting the dump. We will take a look soon.

**Comment by sunilmut** [comment]
We understand the bug. Thanks for the dump.

**Comment by alecbrick** [comment]
Any updates on this? I'm experiencing what I believe to be similar but related issues.

**Comment by ProIcons** [comment]
I'm the original author of the Issue on WSL issue tracker.

Well Subsystem updates only when a new Redstone Update is available unless you are participating on the insider program of microsoft (which is not recommended if you are using that os as your daily driver). 
Having said that the issues with ipfs-update, and ipfs daemon are not present on Windows 1803 (RS4).

Have you updated your Windows Version to RS4 (1803) ? I did i tested out again the ipfs-update which was all the time crashing, and got no crash, i also played around with daemon and didn't faced a crash either. 

That being said, issue is still open on the WSL issue tracker, so it might be partially fixed. 

I tested out the following lines on both windows 1709 and 1803 today 12/5/2018

```bash
wget https://dist.ipfs.io/ipfs-update/v1.5.2/ipfs-update_v1.5.2_linux-amd64.tar.gz
tar -xzf ipfs-update_v1.5.2_linux-amd64.tar.gz
cd ipfs-update
./install.sh
ipfs-update install latest
```

The result on 1803 was the expected
The result on 1708 was a crash.
